src/HotChocolate/Core/src/Subscriptions.Postgres/PostgresChannelWriter.cs

```diff
@@ -93,40 +93,52 @@ private async Task HandleMessage(NpgsqlConnection connection, CancellationToken
             // firstMessage that was already read from the channel
             ct.ThrowIfCancellationRequested();
 
-            await using var batch = connection.CreateBatch();
+            var payloads = new string[messages.Count];
 
-            foreach (var message in messages)
+            for (var i = 0; i < messages.Count; i++)
             {
-                var command = batch.CreateBatchCommand();
+                payloads[i] = messages[i].FormattedPayload;
+            }
+
+            const string sql =
+                """
+                SELECT
+                    pg_notify(t.channel, t.message)
+                from
+                    (select
+                         @channel as channel,
+                         unnest(@messages) as message
+                     ) as t;
+                """;
+
+            await using var command = connection.CreateCommand();
 
-                command.CommandText = "SELECT pg_notify(@channel, @message);";
+            command.CommandText = sql;
 
-                var channel = new NpgsqlParameter("channel", DbType.String)
+            command.Parameters.Add(
+                new NpgsqlParameter("channel", NpgsqlDbType.Text)
                 {
                     Value = _channelName
-                };
-                var msg = new NpgsqlParameter("message", DbType.String)
-                {
-                    Value = message.FormattedPayload
-                };
-                command.Parameters.Add(channel);
-                command.Parameters.Add(msg);
+                });
 
-                batch.BatchCommands.Add(command);
-            }
+            command.Parameters.Add(
+                new NpgsqlParameter("messages", NpgsqlDbType.Array | NpgsqlDbType.Varchar)
+                {
+                    Value = payloads
+                });
 
-            await batch.PrepareAsync(ct);
-            await batch.ExecuteNonQueryAsync(ct);
+            await command.PrepareAsync(ct);
+            await command.ExecuteNonQueryAsync(ct);
         }
         catch (Exception ex)
         {
-            var msg = string.Format(ChannelWriter_FailedToSend, messages.Count, ex.Message);
+            var msg = string.Format(ChannelWriter_FailedToSend, messages.Count, ex.ToString());
             _diagnosticEvents.ProviderInfo(msg);
 
             // if we cannot send the message we put it back into the channel
             foreach (var message in messages)
             {
-                await _channel.Writer.WriteAsync(message, ct);
+                await _channel.Writer.WriteAsync(message, CancellationToken.None);
```

Comment:
@williamdenton  can you add a comment here why you do not pass the cancellation token on purpose?

Reply:
Also, i think we should do a TryWrite here, otherwise we probably run into issues when the channel is full

Reply:
good call, that would be a deadlock situation